### ISSUE TITLE
Make available free space human readable

### ISSUE THREAD

**Issue by mnewt** [Fuco1/dired-hacks#150]
Thanks for the awesome package. Do you want to see the available space in a dired listing to use human readable format, e.g. "100G" instead of "1073741824"?

```emacs-lisp
(defun dired-make-available-human-readable ()
  "Make the available space at the top of the dired buffer human
readable. This seems like a crazy way to do it but as of Emacs
27.1 dired uses `file-system-info', which returns free space in
numerical form. So we wait until dired has inserted the listing
and then modify it."
  (save-excursion
    (goto-char (point-min))
    (let* ((buffer-read-only nil)
           (avail (string-trim (replace-regexp-in-string
                                "Avail" ""
                                (shell-command-to-string "df --output=avail -h .")
                                t t))))
      (when (re-search-forward
             "^\\s-*total used in directory [0-9BGKM]+ available \\([0-9]+\\)"
             nil t)
        (replace-match (propertize avail 'invisible 'dired-hide-details-information)
                       t nil nil 1)))))

(add-hook 'dired-after-readin-hook #'dired-make-available-human-readable)
```

If you want this, feel free to use it in any way you like, or I can submit a pull request if you prefer. Cheers!

**Comment by Fuco1** [comment]
Can you post a screenshot of what this does?  My dired has this:

![](https://i.imgur.com/ii7aNum.png)

**Comment by mnewt** [comment]
It does basically the same thing. Here is an example of my version.

<img width="420" alt="Screen Shot 2019-04-23 at 10 21 21 AM" src="https://user-images.githubusercontent.com/1477258/56602551-9632b380-65b2-11e9-9242-dd70db99e934.png">

I tried your version (Found it [here](https://github.com/Fuco1/.emacs.d/blob/22983a1a97d9c7ebe3e1b38e2f30e045525de5ae/files/dired-defs.el#L273) and there are a few main differences that I can see:

1. My version was just calling out to `df`, whereas you have a pure lisp function formatting the sizes.
2. Your version will replace anything after "directory". I have `(setq dired-listing-switches "-aFhl")` in my config so in that case it mangles an already human readable string.
3. Your version produces an artifact (see below) when switching to `dired-hide-details-mode`. Mine tags the inserted string as invisible when hiding details.

<img width="207" alt="Screen Shot 2019-04-23 at 10 31 54 AM" src="https://user-images.githubusercontent.com/1477258/56602770-12c59200-65b3-11e9-89b1-0f2b246d828b.png">

The best of both seems to be something like this:

```emacs-lisp
(defun dired-format-summary-line ()
  "Format the `total used in directory' and `available' space as
human readable."
  (save-excursion
    (goto-char (point-min))
    (forward-line)
    (let ((inhibit-read-only t)
          (limit (line-end-position)))
      (while (re-search-forward "\\(directory\\|available\\) \\(\\<[0-9]+\\>\\)" nil t)
        (replace-match (save-match-data
                         (propertize (string-trim
                                      (ls-lisp-format-file-size
                                       (* 1024 (string-to-number (match-string 2))) t))
                                     'invisible 'dired-hide-details-information))
                       t nil nil 2)))))
```

That way, it only replaces the string if it's a number without a letter after it, and it tags it as invisible when `dired-hide-details-mode` is on.

**Comment by Fuco1** [comment]
Thanks! this is quite interesting.  I prefer pure lisp for better portability.

I think this would be a good addition to the `dired-hacks.el` file as it really isn't specific to anything else.

I have `dired-hide-details-mode` active but I still see the whole

      total used in directory 178k available 41266M

Is that configurable?  I prefer to see all of it even when the details are collapsed.

**Comment by mnewt** [comment]
Yes, that's `dired-hide-details-hide-information-lines`, which appears to be set to `t` by default

**Comment by dakra** [comment]
I'm not sure where `ls-lisp-format-file-size` is from but there is also `file-size-human-readable` which could be used.

**Comment by Fuco1** [comment]
@mnewt I see, yes, I have `dired-hide-details-hide-information-lines` set to `nil`.

Can we make the code adding the properties conditional on that?

@dakra Hm, I actually have redefined `ls-lisp-format-file-size` in my config, but it comes from `ls-lisp.el` which is a partial elisp implementation of ls. 

What I wanted is to have it show megabytes and not round up to gigabytes if the number is "too big".  Can `file-size-human-readable` do it somehow?  Either way `file-size-human-readable` sounds like a more stable choice.

**Comment by mnewt** [comment]
> I see, yes, I have dired-hide-details-hide-information-lines set to nil.
> Can we make the code adding the properties conditional on that?

Sure but I don't understand why. Dired itself always prints the information lines and just conditionally sets them to visible or invisible. Why not stick with that method? Or maybe I don't follow.

> What I wanted is to have it show megabytes and not round up to gigabytes if the number is "too big". Can file-size-human-readable do it somehow? Either way file-size-human-readable sounds like a more stable choice.

Looking at the source, there is no "level" provision in `file-size-human-readable` to choose to display sizes in KB, MB, or whatever. I guess it's just a judgment call on whether to:

1. Use `file-size-human-readable`, which is in `files.el` so you know it will be already loaded and live with it automatically picking the level
2. Use `ls-lisp-format-file-size` and load `ls-lisp.el`. There is a good chance that would not already be loaded but it ships with Emacs. Seems stable to me. It wasn't previously loaded in my current Emacs session for what that's worth.
3. Copy `ls-lisp-format-file-size` into `dired-hacks`

I don't have a strong feeling but I guess I would just choose option 1 personally just for efficiency's sake and I would think that's what most people would expect.

Could be overthinking this :) But that display in bytes really bugged me until I did something about it!

**Comment by Fuco1** [comment]
The reason I like it in MB is that sometimes I spam `g` (i.e. hold it down) to see if the counter is moving when I'm doing something like downloads if/when I'm not sure if something is going on.  With GBs it would take a while to take effect.

> Sure but I don't understand why. Dired itself always prints the information lines and just conditionally sets them to visible or invisible. Why not stick with that method? Or maybe I don't follow.

If we don't add the condition then it would always hide the numbers if if toggle the hide details mode, right?  Currently I have that on but I see the whole line with the free space

![](https://i.imgur.com/JbAm90H.png)

**Comment by mnewt** [comment]
I hope I understand you correctly.

Dired always renders all the details and then what is visible is controlled by the `buffer-invisibility-spec`, and more specifically the `dired-hide-details-information` tag. As long as the region is has that text property, we can just render it and let dired figure out whether to make it visible based on its settings.

Did you test my snippet? I think it will work for you

**Comment by Fuco1** [comment]
No I didn't test it.  I think this is a good feature so you can open a pull request, then I can fetch that and test it easily.

We should probably put it in the `dired-hacks` file.

**Comment by Fuco1** [comment]
The add/remove hook feature should be wrapped in a minor mode that the user can enable. See for example `dired-collapse` for how to add a minor mode.

**Comment by mnewt** [comment]
Closed in favor of the PR #154
